# D605 Task 3: Solves an Optimization Problem

### A. Using the mathematical representations that you developed in Task 2, write a program in Python to solve the optimization problem computationally

In [1]:
from pulp import *

In [2]:
#hubs with monthly capacities (tons)
hubs = {
    "Cincinnati": 95650,
    "FortWorth": 44350
}

#focus cities with monthly capacities (tons)
focus_cities = {
    "Leipzig": 85000,
    "Hyderabad": 19000,
    "SanBernardino": 36000
}

#fulfillment centers with monthly demand (tons)
fulfillment_centers = {
    "Paris": 6500, "Cologne": 640, "Hanover": 180, "Bangalore": 9100, "Coimbatore": 570,
    "Delhi": 19000, "Mumbai": 14800, "Cagliari": 90, "Catania": 185, "Milan": 800, "Rome": 1700,
    "Katowice": 170, "Barcelona": 2800, "Madrid": 3700, "Castle Donington": 30, "London": 6700,
    "Mobile": 190, "Anchorage": 175, "Fairbanks": 38, "Phoenix": 2400, "Los Angeles": 7200,
    "Ontario": 100, "Riverside": 1200, "Sacramento": 1100, "San Francisco": 1900, "Stockton": 240,
    "Denver": 1500, "Hartford": 540, "Miami": 3400, "Lakeland": 185, "Tampa": 1600, "Atlanta": 3000,
    "Honolulu": 500, "Kahului/Maui": 16, "Kona": 63, "Chicago": 5100, "Rockford": 172,
    "Fort Wayne": 200, "South Bend": 173, "Des Moines": 300, "Wichita": 290, "New Orleans": 550,
    "Baltimore": 1300, "Minneapolis": 1700, "Kansas City": 975, "St. Louis": 1200, "Omaha": 480,
    "Manchester": 100, "Albuquerque": 450, "New York": 11200, "Charlotte": 900, "Toledo": 290,
    "Wilmington": 150, "Portland": 1200, "Allentown": 420, "Pittsburgh": 1000, "San Juan": 1100,
    "Nashville": 650, "Austin": 975, "Dallas": 3300, "Houston": 3300, "San Antonio": 1100,
    "Richmond": 600, "Seattle/Tacoma": 2000, "Spokane": 260
}

<p>In this code chunk, I created three dictionaries. First is "hubs" and it defines the monthly capacity for each hub. Second is "focus_cities" and it defines the monthly capacity for each focus city. Third is "fulfillment centers" and it shows the monthly demand for each fulfillment center.</p>

In [3]:
cost_data = {
    #hubs to focus cities
    ("Cincinnati", "Leipzig"): 1.5,
    ("Cincinnati", "SanBernardino"): 0.5,
    ("FortWorth", "SanBernardino"): 0.5,

    #hubs to fulfillment centers
    ("Cincinnati", "Paris"): 1.6, ("Cincinnati", "Cologne"): 1.5, ("Cincinnati", "Hanover"): 1.5,
    ("Cincinnati", "Cagliari"): 1.5, ("Cincinnati", "Catania"): 1.5, ("Cincinnati", "Milan"): 1.5,
    ("Cincinnati", "Rome"): 1.5, ("Cincinnati", "Katowice"): 1.4, ("Cincinnati", "Barcelona"): 1.5,
    ("Cincinnati", "Madrid"): 1.6, ("Cincinnati", "Castle Donington"): 1.4, ("Cincinnati", "London"): 1.6,
    ("Cincinnati", "Mobile"): 0.5, ("Cincinnati", "Anchorage"): 1.3, ("Cincinnati", "Fairbanks"): 1.4,
    ("Cincinnati", "Phoenix"): 0.5, ("Cincinnati", "Los Angeles"): 0.5, ("Cincinnati", "Ontario"): 0.5,
    ("Cincinnati", "Riverside"): 0.5, ("Cincinnati", "Sacramento"): 0.5, ("Cincinnati", "San Francisco"): 0.5,
    ("Cincinnati", "Stockton"): 0.5, ("Cincinnati", "Denver"): 0.5, ("Cincinnati", "Hartford"): 0.5,
    ("Cincinnati", "Miami"): 0.5, ("Cincinnati", "Lakeland"): 0.5, ("Cincinnati", "Tampa"): 0.5,
    ("Cincinnati", "Atlanta"): 0.5, ("Cincinnati", "Chicago"): 0.5, ("Cincinnati", "Rockford"): 0.5,
    ("Cincinnati", "Fort Wayne"): 0.5, ("Cincinnati", "South Bend"): 0.5, ("Cincinnati", "Des Moines"): 0.5,
    ("Cincinnati", "Wichita"): 0.5, ("Cincinnati", "New Orleans"): 0.5, ("Cincinnati", "Baltimore"): 0.5,
    ("Cincinnati", "Minneapolis"): 0.5, ("Cincinnati", "Kansas City"): 0.5, ("Cincinnati", "St. Louis"): 0.5,
    ("Cincinnati", "Omaha"): 0.5, ("Cincinnati", "Manchester"): 0.5, ("Cincinnati", "Albuquerque"): 0.5,
    ("Cincinnati", "New York"): 0.5, ("Cincinnati", "Charlotte"): 0.5, ("Cincinnati", "Toledo"): 0.5,
    ("Cincinnati", "Wilmington"): 0.5, ("Cincinnati", "Portland"): 0.5, ("Cincinnati", "Allentown"): 0.5,
    ("Cincinnati", "Pittsburgh"): 0.5, ("Cincinnati", "San Juan"): 0.5, ("Cincinnati", "Nashville"): 0.5,
    ("Cincinnati", "Austin"): 0.5, ("Cincinnati", "Dallas"): 0.5, ("Cincinnati", "Houston"): 0.5,
    ("Cincinnati", "San Antonio"): 0.5, ("Cincinnati", "Richmond"): 0.5, ("Cincinnati", "Seattle/Tacoma"): 0.5,
    ("Cincinnati", "Spokane"): 0.5, ("FortWorth", "Mobile"): 0.5, ("FortWorth", "Anchorage"): 1.0,
    ("FortWorth", "Fairbanks"): 1.0, ("FortWorth", "Phoenix"): 0.5, ("FortWorth", "Los Angeles"): 0.5,
    ("FortWorth", "Ontario"): 0.5, ("FortWorth", "Riverside"): 0.5, ("FortWorth", "Sacramento"): 0.5,
    ("FortWorth", "San Francisco"): 0.5, ("FortWorth", "Stockton"): 0.5, ("FortWorth", "Denver"): 0.5,
    ("FortWorth", "Hartford"): 0.5, ("FortWorth", "Miami"): 0.5, ("FortWorth", "Lakeland"): 0.5,
    ("FortWorth", "Tampa"): 0.5, ("FortWorth", "Atlanta"): 0.5, ("FortWorth", "Honolulu"): 0.5,
    ("FortWorth", "Kahului/Maui"): 0.5, ("FortWorth", "Kona"): 0.5, ("FortWorth", "Chicago"): 0.5,
    ("FortWorth", "Rockford"): 0.5, ("FortWorth", "Fort Wayne"): 0.5, ("FortWorth", "South Bend"): 0.5,
    ("FortWorth", "Des Moines"): 0.5, ("FortWorth", "Wichita"): 0.5, ("FortWorth", "New Orleans"): 0.5,
    ("FortWorth", "Baltimore"): 0.5, ("FortWorth", "Minneapolis"): 0.5, ("FortWorth", "Kansas City"): 0.5,
    ("FortWorth", "St. Louis"): 0.5, ("FortWorth", "Omaha"): 0.5, ("FortWorth", "Manchester"): 0.5,
    ("FortWorth", "Albuquerque"): 0.5, ("FortWorth", "New York"): 0.5, ("FortWorth", "Charlotte"): 0.5,
    ("FortWorth", "Toledo"): 0.5, ("FortWorth", "Wilmington"): 0.5, ("FortWorth", "Portland"): 0.5,
    ("FortWorth", "Allentown"): 0.5, ("FortWorth", "Pittsburgh"): 0.5, ("FortWorth", "San Juan"): 0.5,
    ("FortWorth", "Nashville"): 0.5, ("FortWorth", "Austin"): 0.25, ("FortWorth", "Houston"): 0.25,
    ("FortWorth", "San Antonio"): 0.25, ("FortWorth", "Richmond"): 0.5, ("FortWorth", "Seattle/Tacoma"): 0.5,
    ("FortWorth", "Spokane"): 0.5,

    #focus cities to fulfillment centers
    ("Leipzig", "Paris"): 0.5, ("Leipzig", "Cologne"): 0.5, ("Leipzig", "Hanover"): 0.5,
    ("Leipzig", "Cagliari"): 0.5, ("Leipzig", "Catania"): 0.5, ("Leipzig", "Milan"): 0.5,
    ("Leipzig", "Rome"): 0.5, ("Leipzig", "Katowice"): 0.5, ("Leipzig", "Barcelona"): 0.5,
    ("Leipzig", "Madrid"): 0.5, ("Leipzig", "Castle Donington"): 0.5, ("Leipzig", "London"): 0.75,
    ("Leipzig", "Hartford"): 1.5, ("Leipzig", "Baltimore"): 1.5, ("Leipzig", "Manchester"): 1.5,
    ("Leipzig", "New York"): 1.6, ("Leipzig", "Allentown"): 1.5, ("Leipzig", "Bangalore"): 1.5,
    ("Leipzig", "Coimbatore"): 1.5, ("Leipzig", "Delhi"): 1.5, ("Leipzig", "Mumbai"): 1.5,
    ("Hyderabad", "Paris"): 1.1, ("Hyderabad", "Cologne"): 1.0, ("Hyderabad", "Hanover"): 1.0,
    ("Hyderabad", "Cagliari"): 1.0, ("Hyderabad", "Catania"): 1.0, ("Hyderabad", "Milan"): 1.0,
    ("Hyderabad", "Rome"): 1.1, ("Hyderabad", "Katowice"): 1.0, ("Hyderabad", "Barcelona"): 1.0,
    ("Hyderabad", "Madrid"): 1.1, ("Hyderabad", "London"): 1.1,
    ("Hyderabad", "Bangalore"): 0.5, ("Hyderabad", "Coimbatore"): 0.5, ("Hyderabad", "Delhi"): 0.5,
    ("Hyderabad", "Mumbai"): 0.5, ("SanBernardino", "Mobile"): 0.5, ("SanBernardino", "Anchorage"): 0.7,
    ("SanBernardino", "Fairbanks"): 0.7, ("SanBernardino", "Phoenix"): 0.5, ("SanBernardino", "Sacramento"): 0.5,
    ("SanBernardino", "San Francisco"): 0.5, ("SanBernardino", "Stockton"): 0.5, ("SanBernardino", "Denver"): 0.5,
    ("SanBernardino", "Hartford"): 0.5, ("SanBernardino", "Miami"): 0.7, ("SanBernardino", "Lakeland"): 0.7,
    ("SanBernardino", "Tampa"): 0.7, ("SanBernardino", "Atlanta"): 0.6, ("SanBernardino", "Honolulu"): 0.5,
    ("SanBernardino", "Kahului/Maui"): 0.5, ("SanBernardino", "Kona"): 0.5, ("SanBernardino", "Chicago"): 0.5,
    ("SanBernardino", "Rockford"): 0.5, ("SanBernardino", "Fort Wayne"): 0.5, ("SanBernardino", "South Bend"): 0.5,
    ("SanBernardino", "Des Moines"): 0.5, ("SanBernardino", "Wichita"): 0.5, ("SanBernardino", "New Orleans"): 0.5,
    ("SanBernardino", "Baltimore"): 0.7, ("SanBernardino", "Minneapolis"): 0.5, ("SanBernardino", "Kansas City"): 0.5,
    ("SanBernardino", "St. Louis"): 0.5, ("SanBernardino", "Omaha"): 0.5, ("SanBernardino", "Manchester"): 0.7,
    ("SanBernardino", "Albuquerque"): 0.5, ("SanBernardino", "New York"): 0.7, ("SanBernardino", "Charlotte"): 0.7,
    ("SanBernardino", "Toledo"): 0.5, ("SanBernardino", "Wilmington"): 0.7, ("SanBernardino", "Portland"): 0.5,
    ("SanBernardino", "Allentown"): 0.7, ("SanBernardino", "Pittsburgh"): 0.6, ("SanBernardino", "San Juan"): 1.0, 
    ("SanBernardino", "Nashville"): 0.5, ("SanBernardino", "Austin"): 0.5, ("SanBernardino", "Dallas"): 0.5,
    ("SanBernardino", "Houston"): 0.5, ("SanBernardino", "San Antonio"): 0.5, ("SanBernardino", "Richmond"): 0.7,
    ("SanBernardino", "Seattle/Tacoma"): 0.5, ("SanBernardino", "Spokane"): 0.5
}

<p>In this code chunk, I created a dictionary called "cost_data" that stores shipping costs for each route. Each key is a route pair (origin, desitination), and each value is the cost per ton. I grouped the data into three types of routes: from hubs to focus cities, from hubs to fulfillment centers, and from focus cities to fulfillment centers.</p>

In [4]:
model = LpProblem("Amazon_Air_Optimization", LpMinimize)

#dividing cost_data by type
cost_hub_to_focus = {(i, j): c for (i, j), c in cost_data.items() if i in hubs and j in focus_cities}
cost_hub_to_center = {(i, k): c for (i, k), c in cost_data.items() if i in hubs and k in fulfillment_centers}
cost_focus_to_center = {(j, k): c for (j, k), c in cost_data.items() if j in focus_cities and k in fulfillment_centers}

#defining decision variables
x = LpVariable.dicts("x", cost_hub_to_focus, lowBound=0, cat="Continuous")
y = LpVariable.dicts("y", cost_hub_to_center, lowBound=0, cat="Continuous")
z = LpVariable.dicts("z", cost_focus_to_center, lowBound=0, cat="Continuous")

#creating the objective function, which is to minimize total transportation cost
model += (
    lpSum(cost_hub_to_focus[i, j] * x[i, j] for (i, j) in x) +
    lpSum(cost_hub_to_center[i, k] * y[i, k] for (i, k) in y) +
    lpSum(cost_focus_to_center[j, k] * z[j, k] for (j, k) in z)
)

<p>In this code chunk, I started by creating a new linear programming problem called "Amazon_Air_Optimization". I used the argument "LpMinimize" to specify that the goal is to minimize the objective function. Next, I split the "cost_data" dictionary based on route types to make the model easier to work with. The variable "cost_hub_to_focus" contains costs for routes from hubs to focus cities. The variable "cost_hub_to_center" contains costs for routes directly from hubs to fulfillment centers. The variable "cost_focus_to_center" contains costs for routes from focus cities to fulfillment centers.</p>

<p>Then, I created three decision variables that represent shipment quantities for each route type. The variable "x[i, j]" is the quantity shipped from hub i to focus city j. The variable "y[i, k]" is the quantity shipped directly from hub i to fulfillment center k. The variable "z[j, k]" is the quantity shipped from focus city j to fulfillment center k. I made sure that the variables are continuous and that there are no negative shipment quantities. Finally, I created the objective function to minimize the total transportation cost. This was done by multiplying each shipment quantity by its cost and summing all route costs.</p>

In [5]:
#hub capacity constraint
for i in hubs:
    model += (
        lpSum(x[i, j] for j in focus_cities if (i, j) in x) +
        lpSum(y[i, k] for k in fulfillment_centers if (i, k) in y)
    ) <= hubs[i]

#focus city capacity constraint
for j in focus_cities:
    model += (
        lpSum(x[i, j] for i in hubs if (i, j) in x)
    ) <= focus_cities[j]

#flow balance at focus cities constraint
for j in focus_cities:
    model += (
        lpSum(x[i, j] for i in hubs if (i, j) in x)
        == lpSum(z[j, k] for k in fulfillment_centers if (j, k) in z)
    )

#fulfillment center demand constraint
for k in fulfillment_centers:
    model += (
        lpSum(y[i, k] for i in hubs if (i, k) in y) +
        lpSum(z[j, k] for j in focus_cities if (j, k) in z)
    ) == fulfillment_centers[k]

<p>In this code chunk, I defined the constraints for the optimization problem. These constraints make sure the solution follows real-world limits like capacity and demand. The first constraint is the hub capacity constraint, which ensures that the total cargo shipped out from each hub does not exceed its monthly capacity. For example, the Cincinnati hub cannot ship more than 95,650 tons per month to focus cities and fulfillment centers combined. The second constraint is the focus city capacity constraint, which ensures that the total cargo entering each focus city stays within its handling capacity. For example, the focus city Leipzig can receive up to 85,000 tons from all hubs each month. The third constraint is the flow balance at focus cities, which ensures that all cargo received by a focus city is sent to fulfillment centers without any surplus or shortage. For example, if the focus city Leipzig receives 50,000 tons, it must send exactly 50,000 tons to the fulfillment centers. The fourth constraint is the fulfillment center demand constraint, which ensures each fulfillment center receives exactly the amount of cargo it needs. For example, the Paris fulfillment center must receive 6,500 tons per month, either directly from hubs or through focus cities.</p>

In [6]:
model.solve()

#displaying results
print("Status:", LpStatus[model.status])
print("Total Cost: $", round(value(model.objective), 2))
print("\nRoutes Used:")
for var in model.variables():
    if var.varValue > 0:
        print(f"{var.name} = {var.varValue}")

Status: Optimal
Total Cost: $ 199476.25

Routes Used:
x_('Cincinnati',_'Leipzig') = 43470.0
y_('Cincinnati',_'Atlanta') = 3000.0
y_('Cincinnati',_'Baltimore') = 1300.0
y_('Cincinnati',_'Barcelona') = 2800.0
y_('Cincinnati',_'Cagliari') = 90.0
y_('Cincinnati',_'Castle_Donington') = 30.0
y_('Cincinnati',_'Catania') = 185.0
y_('Cincinnati',_'Charlotte') = 900.0
y_('Cincinnati',_'Chicago') = 5100.0
y_('Cincinnati',_'Cologne') = 640.0
y_('Cincinnati',_'Dallas') = 3300.0
y_('Cincinnati',_'Denver') = 1500.0
y_('Cincinnati',_'Hanover') = 180.0
y_('Cincinnati',_'Hartford') = 540.0
y_('Cincinnati',_'Kansas_City') = 975.0
y_('Cincinnati',_'Katowice') = 170.0
y_('Cincinnati',_'Lakeland') = 185.0
y_('Cincinnati',_'London') = 6700.0
y_('Cincinnati',_'Madrid') = 3700.0
y_('Cincinnati',_'Miami') = 3400.0
y_('Cincinnati',_'Milan') = 800.0
y_('Cincinnati',_'New_Orleans') = 550.0
y_('Cincinnati',_'Omaha') = 480.0
y_('Cincinnati',_'Paris') = 6500.0
y_('Cincinnati',_'Phoenix') = 2400.0
y_('Cincinnati',_'Po

<p>In this code chunk, I solved the optimization problem and displayed the results. First, I used PuLP's default CBC solver to find the solution that minimizes total transportation cost while meeting all constraints. Then, I printed the solution status to check if the solver found an optimal result. After that, I printed the total minimimzed cost. Lastly, I printed only the routes that were used, which are the ones with shipment quantities greater than zero.</p>

<p>Looking at the output, the status is optimal, which means the solver found a solution that meets all constraints and gives the lowest possible total transportation cost. The total cost is $199,476.25. All the active routes and their shipment amounts are listed. For example, the route "x_('Cincinnati',_'Leipzig') = 43470.0" shows that the Cincinnati hub shipped 43,470 tons to the Leipzig focus city, which likely forwards those shipments to fulfillment centers. Another example is "y_('Cincinnati',_'Atlanta') = 3000.0", which means 3,000 tons were shipped directly from the Cincinnati hub to the Atlanta fulfillment center. Lastly, "z_('Leipzig',_'Delhi') = 19000.0" shows that the Leipzig focus city sent 19,000 tons to the Delhi fulfillment center.</p>

<p>There are no routes from the San Bernardino and Hyderabad focus cities, which means the solver chose not to use them in the optimal solution. This likely happened for a few reasons. First, shipping directly from the Cincinnati and Fort Worth hubs or through Leipzig was cheaper than using San Bernardino or Hyderabad. Second, all fulfillment center demands and capacity limits were met without needing those two cities. Finally, because the goal was to minimize total transportation cost, the solver left out any routes that were not needed or cost-effective.</p>

### A1. Demonstrate that the solver provided a solution to the optimization problem

In [7]:
model.solve()

#displaying results
print("Status:", LpStatus[model.status])
print("Total Cost: $", round(value(model.objective), 2))

Status: Optimal
Total Cost: $ 199476.25


<p>The output shows that the solver status is “optimal”, which means the optimization problem was solved successfully and all constraints were met. The total transportation cost is $199,476.25, which is the lowest cost needed to meet all demand while staying within hub and focus city capacities.</p>

In [8]:
print("\nRoutes Used:")
for var in model.variables():
    if var.varValue > 0:
        print(f"{var.name} = {var.varValue}")


Routes Used:
x_('Cincinnati',_'Leipzig') = 43470.0
y_('Cincinnati',_'Atlanta') = 3000.0
y_('Cincinnati',_'Baltimore') = 1300.0
y_('Cincinnati',_'Barcelona') = 2800.0
y_('Cincinnati',_'Cagliari') = 90.0
y_('Cincinnati',_'Castle_Donington') = 30.0
y_('Cincinnati',_'Catania') = 185.0
y_('Cincinnati',_'Charlotte') = 900.0
y_('Cincinnati',_'Chicago') = 5100.0
y_('Cincinnati',_'Cologne') = 640.0
y_('Cincinnati',_'Dallas') = 3300.0
y_('Cincinnati',_'Denver') = 1500.0
y_('Cincinnati',_'Hanover') = 180.0
y_('Cincinnati',_'Hartford') = 540.0
y_('Cincinnati',_'Kansas_City') = 975.0
y_('Cincinnati',_'Katowice') = 170.0
y_('Cincinnati',_'Lakeland') = 185.0
y_('Cincinnati',_'London') = 6700.0
y_('Cincinnati',_'Madrid') = 3700.0
y_('Cincinnati',_'Miami') = 3400.0
y_('Cincinnati',_'Milan') = 800.0
y_('Cincinnati',_'New_Orleans') = 550.0
y_('Cincinnati',_'Omaha') = 480.0
y_('Cincinnati',_'Paris') = 6500.0
y_('Cincinnati',_'Phoenix') = 2400.0
y_('Cincinnati',_'Portland') = 1200.0
y_('Cincinnati',_'Rich

<p>The routes listed above show the shipment amounts for each route used in the optimal solution. For example, the route "x_('Cincinnati',_'Leipzig') = 43470.0" shows that the Cincinnati hub shipped 43,470 tons to the Leipzig focus city, which likely forwarded those shipments to fulfillment centers. Another example is "y_('Cincinnati',_'Atlanta') = 3000.0", which means 3,000 tons were shipped directly from the Cincinnati hub to the Atlanta fulfillment center. Lastly, "z_('Leipzig',_'Delhi') = 19000.0" shows that the Leipzig focus city sent 19,000 tons to the Delhi fulfillment center. This output confirms that the solver assigned values to the decision variables while meeting all constraints.</p>

### B. Analyze the output of your model to demonstrate that the solution satisfies the requirements of the problem

<p>The goal of this optimization problem was to minimize the total transportation cost while making sure all fulfillment centers receive the cargo they need and that hubs and focus cities stay within their capacity limits.</p>

In [9]:
model.solve()

#displaying results
print("Status:", LpStatus[model.status])
print("Total Cost: $", round(value(model.objective), 2))

Status: Optimal
Total Cost: $ 199476.25


<p>The solver returned a status of ”optimal”, which means the optimization problem was solved successfully and all constraints were met. The total transportation cost is $199,476.25, which is the lowest cost needed to meet all demand while staying within hub and focus city capacities.</p>

In [10]:
print("\nRoutes Used:")
for var in model.variables():
    if var.varValue > 0:
        print(f"{var.name} = {var.varValue}")


Routes Used:
x_('Cincinnati',_'Leipzig') = 43470.0
y_('Cincinnati',_'Atlanta') = 3000.0
y_('Cincinnati',_'Baltimore') = 1300.0
y_('Cincinnati',_'Barcelona') = 2800.0
y_('Cincinnati',_'Cagliari') = 90.0
y_('Cincinnati',_'Castle_Donington') = 30.0
y_('Cincinnati',_'Catania') = 185.0
y_('Cincinnati',_'Charlotte') = 900.0
y_('Cincinnati',_'Chicago') = 5100.0
y_('Cincinnati',_'Cologne') = 640.0
y_('Cincinnati',_'Dallas') = 3300.0
y_('Cincinnati',_'Denver') = 1500.0
y_('Cincinnati',_'Hanover') = 180.0
y_('Cincinnati',_'Hartford') = 540.0
y_('Cincinnati',_'Kansas_City') = 975.0
y_('Cincinnati',_'Katowice') = 170.0
y_('Cincinnati',_'Lakeland') = 185.0
y_('Cincinnati',_'London') = 6700.0
y_('Cincinnati',_'Madrid') = 3700.0
y_('Cincinnati',_'Miami') = 3400.0
y_('Cincinnati',_'Milan') = 800.0
y_('Cincinnati',_'New_Orleans') = 550.0
y_('Cincinnati',_'Omaha') = 480.0
y_('Cincinnati',_'Paris') = 6500.0
y_('Cincinnati',_'Phoenix') = 2400.0
y_('Cincinnati',_'Portland') = 1200.0
y_('Cincinnati',_'Rich

<p>The solution satisfies all four constraints. For the hub capacity constraint, both Cincinnati and Fort Worth hubs shipped less than their maximum limits. Looking at the output, the Cincinnati hub shipped a total of 95,518 tons, which is 132 tons under its 95,560-ton capacity. The Fort Worth hub shipped 38,229 tons, staying 6,121 tons below its 44,350-ton capacity.</p>

<p>For the focus city capacity constraint, each focus city received only as much as it could handle. For example, the route "x_('Cincinnati',_'Leipzig') = 43470.0" shows that the Leipzig focus city only received 43,470 tons, which is well within its 85,000-ton capacity.</p>

<p>For the flow balance constraint at focus cities, the total amount received by a focus city equals the amount it sent out. In the output, the Leipzig focus city received 43,470 tons and sent out the same total amount to the fulfillment centers. It shipped 9,100 tons to Bangalore, 570 tons to Coimbatore, 19,000 tons to Delhi, and 14,800 tons to Mumbai. This shows that there was no leftover cargo or shortage at the Leipzig focus city.</p>

<p>For the fulfillment center demand constraint, every fulfillment center received exactly what it needed. For example, the route “z_('Leipzig',_'Delhi') = 19000.0” shows that the Delhi fulfillment center received its full demand from the Leipzig focus city, and the route “y_('Cincinnati',_'Atlanta') = 3000.0” shows that the Atlanta fulfillment center received 3,000 tons directly from the Cincinnati hub.</p>

<p>There are no routes from the San Bernardino and Hyderabad focus cities, which means the solver chose not to use them in the optimal solution. This likely happened for a few reasons. First, shipping directly from the Cincinnati and Fort Worth hubs or through Leipzig was cheaper than using San Bernardino or Hyderabad. Second, all fulfillment center demands and capacity limits were met without needing those two cities. Finally, because the goal was to minimize total transportation cost, the solver left out any routes that were not needed or cost-effective.</p>

### B1. Demonstrate that the constraints of the optimization problem are satisfied

In [11]:
model.solve()

#displaying results
print("Status:", LpStatus[model.status])
print("Total Cost: $", round(value(model.objective), 2))

Status: Optimal
Total Cost: $ 199476.25


<p>The solver returned an “optimal” status, which means it found a feasible solution that satisfies all constraints, including hub capacity, focus city capacity, flow balance at focus cities, and fulfillment center demand. The total transportation cost is $199,476.25, which is the lowest cost needed to meet all demand while staying within hub and focus city capacities.</p>

In [12]:
print("\nRoutes Used:")
for var in model.variables():
    if var.varValue > 0:
        print(f"{var.name} = {var.varValue}")


Routes Used:
x_('Cincinnati',_'Leipzig') = 43470.0
y_('Cincinnati',_'Atlanta') = 3000.0
y_('Cincinnati',_'Baltimore') = 1300.0
y_('Cincinnati',_'Barcelona') = 2800.0
y_('Cincinnati',_'Cagliari') = 90.0
y_('Cincinnati',_'Castle_Donington') = 30.0
y_('Cincinnati',_'Catania') = 185.0
y_('Cincinnati',_'Charlotte') = 900.0
y_('Cincinnati',_'Chicago') = 5100.0
y_('Cincinnati',_'Cologne') = 640.0
y_('Cincinnati',_'Dallas') = 3300.0
y_('Cincinnati',_'Denver') = 1500.0
y_('Cincinnati',_'Hanover') = 180.0
y_('Cincinnati',_'Hartford') = 540.0
y_('Cincinnati',_'Kansas_City') = 975.0
y_('Cincinnati',_'Katowice') = 170.0
y_('Cincinnati',_'Lakeland') = 185.0
y_('Cincinnati',_'London') = 6700.0
y_('Cincinnati',_'Madrid') = 3700.0
y_('Cincinnati',_'Miami') = 3400.0
y_('Cincinnati',_'Milan') = 800.0
y_('Cincinnati',_'New_Orleans') = 550.0
y_('Cincinnati',_'Omaha') = 480.0
y_('Cincinnati',_'Paris') = 6500.0
y_('Cincinnati',_'Phoenix') = 2400.0
y_('Cincinnati',_'Portland') = 1200.0
y_('Cincinnati',_'Rich

<p>The solution satisfies all four constraints. For the hub capacity constraint, both Cincinnati and Fort Worth hubs shipped less than their maximum limits. Looking at the output, the Cincinnati hub shipped a total of 95,518 tons, which is 132 tons under its 95,560-ton capacity. The Fort Worth hub shipped 38,229 tons, staying 6,121 tons below its 44,350-ton capacity.</p>

<p>For the focus city capacity constraint, each focus city received only as much as it could handle. For example, the route "x_('Cincinnati',_'Leipzig') = 43470.0" shows that the Leipzig focus city only received 43,470 tons, which is well within its 85,000-ton capacity.</p>

<p>For the flow balance constraint at focus cities, the total amount received by a focus city equals the amount it sent out. In the output, the Leipzig focus city received 43,470 tons and sent out the same total amount to the fulfillment centers. It shipped 9,100 tons to Bangalore, 570 tons to Coimbatore, 19,000 tons to Delhi, and 14,800 tons to Mumbai. This shows that there was no leftover cargo or shortage at the Leipzig focus city.</p>

<p>For the fulfillment center demand constraint, every fulfillment center received exactly what it needed. For example, the route “z_('Leipzig',_'Delhi') = 19000.0” shows that the Delhi fulfillment center received its full demand from the Leipzig focus city, and the route “y_('Cincinnati',_'Atlanta') = 3000.0” shows that the Atlanta fulfillment center received 3,000 tons directly from the Cincinnati hub.</p>

### B2. Demonstrate that the solution includes decision variables, constraints, and the objective function

<p>The optimization model that I created includes decision variables, constraints, and the objective function.</p>

In [13]:
model = LpProblem("Amazon_Air_Optimization", LpMinimize)

#dividing cost_data by type
cost_hub_to_focus = {(i, j): c for (i, j), c in cost_data.items() if i in hubs and j in focus_cities}
cost_hub_to_center = {(i, k): c for (i, k), c in cost_data.items() if i in hubs and k in fulfillment_centers}
cost_focus_to_center = {(j, k): c for (j, k), c in cost_data.items() if j in focus_cities and k in fulfillment_centers}

#defining decision variables
x = LpVariable.dicts("x", cost_hub_to_focus, lowBound=0, cat="Continuous")
y = LpVariable.dicts("y", cost_hub_to_center, lowBound=0, cat="Continuous")
z = LpVariable.dicts("z", cost_focus_to_center, lowBound=0, cat="Continuous")

<p>Here, I started by creating a new linear programming problem called "Amazon_Air_Optimization". I used the argument "LpMinimize" to specify that the goal is to minimize the objective function. Next, I split the "cost_data" dictionary based on route types to make the model easier to work with. The variable "cost_hub_to_focus" contains costs for routes from hubs to focus cities. The variable "cost_hub_to_center" contains costs for routes directly from hubs to fulfillment centers. The variable "cost_focus_to_center" contains costs for routes from focus cities to fulfillment centers.</p>

<p>Then, I created three decision variables that represent shipment quantities for each route type. The variable "x[i, j]" is the quantity shipped from hub i to focus city j. The variable "y[i, k]" is the quantity shipped directly from hub i to fulfillment center k. The variable "z[j, k]" is the quantity shipped from focus city j to fulfillment center k. I made sure that the variables are continuous and that there are no negative shipment quantities.</p>

In [ ]:
#hub capacity constraint
for i in hubs:
    model += (
        lpSum(x[i, j] for j in focus_cities if (i, j) in x) +
        lpSum(y[i, k] for k in fulfillment_centers if (i, k) in y)
    ) <= hubs[i]

#focus city capacity constraint
for j in focus_cities:
    model += (
        lpSum(x[i, j] for i in hubs if (i, j) in x)
    ) <= focus_cities[j]

#flow balance at focus cities constraint
for j in focus_cities:
    model += (
        lpSum(x[i, j] for i in hubs if (i, j) in x)
        == lpSum(z[j, k] for k in fulfillment_centers if (j, k) in z)
    )

#fulfillment center demand constraint
for k in fulfillment_centers:
    model += (
        lpSum(y[i, k] for i in hubs if (i, k) in y) +
        lpSum(z[j, k] for j in focus_cities if (j, k) in z)
    ) == fulfillment_centers[k]

<p>Here, I defined the constraints for the optimization problem. These constraints make sure the solution follows real-world limits like capacity and demand. The first constraint is the hub capacity constraint, which ensures that the total cargo shipped out from each hub does not exceed its monthly capacity. For example, the Cincinnati hub cannot ship more than 95,650 tons per month to focus cities and fulfillment centers combined. The second constraint is the focus city capacity constraint, which ensures that the total cargo entering each focus city stays within its handling capacity. For example, the focus city Leipzig can receive up to 85,000 tons from all hubs each month. The third constraint is the flow balance at focus cities, which ensures that all cargo received by a focus city is sent to fulfillment centers without any surplus or shortage. For example, if the focus city Leipzig receives 50,000 tons, it must send exactly 50,000 tons to the fulfillment centers. The fourth constraint is the fulfillment center demand constraint, which ensures each fulfillment center receives exactly the amount of cargo it needs. For example, the Paris fulfillment center must receive 6,500 tons per month, either directly from hubs or through focus cities.</p>

In [ ]:
#creating the objective function, which is to minimize total transportation cost
model += (
    lpSum(cost_hub_to_focus[i, j] * x[i, j] for (i, j) in x) +
    lpSum(cost_hub_to_center[i, k] * y[i, k] for (i, k) in y) +
    lpSum(cost_focus_to_center[j, k] * z[j, k] for (j, k) in z)
)

<p>Here, I defined the objective function to minimize the total transportation cost. This is done by multiplying the cost per ton for each route by the amount shipped on that route, and then summing these values across all routes.</p>

### B3. Explain why the solution that your code outputs matches your expected output

<p>The solution produced by my code matches my expected output because the model was designed to minimize the total transportation cost while satisfying all constraints. The solver returned an “optimal” status, which confirms that it found a valid and cost-effective solution. This indicates that the model successfully minimized the total transportation cost while meeting every constraint. The route output also shows that all four constraints were satisfied. For the hub capacity constraint, both the Cincinnati and Fort Worth hubs shipped less than their maximum limits. For the focus city capacity constraint, each focus city received only as much as it could handle. For the flow balance constraint at focus cities, the total amount received by each focus city equaled the amount it sent out. For the fulfillment center demand constraint, each fulfillment center received exactly what it needed. The solver also avoided using more expensive focus cities like San Bernardino and Hyderabad, which shows that it chose the most cost-effective routes. This all aligns with the goals and expectations of the model.</p>

### C. Provide a reflection on how the development of your approach matched your expectations of the process

<p>When I first started this task, it took me a while to enter all the hubs, focus cities, and fulfillment centers into the program because there were so many. I also made a few typos in the route data, which I didn’t notice until I ran the program and the solver returned an “infeasible” status, which means the model “has no solution that satisfies all the constraints” (Hofmann, 2021, par. 1). Aside from that, everything else went smoothly. Building the optimization model was fairly straightforward, which I expected since I had already defined the objective function, constraints, and decision variables in Task 2. I did face some challenges in organizing the data, especially when I accidentally placed decision variables in the wrong parts of the formulas.</p>

<p>I was surprised by how efficiently the solver found the most cost-effective routes. I also didn’t expect that some focus cities, like San Bernardino and Hyderabad, would be completely unused, but it made sense once I saw the results. The solver naturally avoided more expensive or unnecessary routes.</p>

<p>Overall, the development process matched my expectations in both structure and outcome. It also reminded me how important it is to carefully check the code and formulas.</p>

### D. Acknowledge sources, using in-text citations and references, for content that is quoted, paraphrased, or summarized

<pre>
Hofmann, F. (2021, March). <i>Trouble shooting - Infeasible Model.</i> Linopy. https://linopy.readthedocs.io/en/latest/infeasible-model.html
</pre>